tuf/testutils/swizzler.go

```diff
@@ -0,0 +1,551 @@
+package testutils
+
+import (
+	"bytes"
+	"path"
+	"time"
+
+	"github.com/docker/notary/cryptoservice"
+	"github.com/docker/notary/passphrase"
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/tuf/data"
+	"github.com/docker/notary/tuf/signed"
+	"github.com/docker/notary/tuf/store"
+	"github.com/jfrazelle/go/canonical/json"
+)
+
+const (
+	maxSize = 5 << 20
+)
+
+// ErrNoKeyForRole returns an error when the cryptoservice provided to
+// MetadataSwizzler has no key for a particular role
+type ErrNoKeyForRole struct {
+	Role string
+}
+
+func (e ErrNoKeyForRole) Error() string {
+	return "Swizzler's cryptoservice has no key for role " + e.Role
+}
+
+// MetadataSwizzler fuzzes the metadata in a MetadataStore
+type MetadataSwizzler struct {
+	gun           string
+	MetadataCache store.MetadataStore
+	CryptoService signed.CryptoService
+	Roles         []string // list of Roles in the metadataStore
+}
+
+// signs the new metadata, replacing whatever signature was there
+func serializeMetadata(cs signed.CryptoService, s *data.Signed, role string,
+	pubKeys ...data.PublicKey) ([]byte, error) {
+
+	// delete the existing signatures
+	s.Signatures = []data.Signature{}
+
+	if len(pubKeys) > 0 {
+		if err := signed.Sign(cs, s, pubKeys...); err != nil {
+			if _, ok := err.(signed.ErrNoKeys); ok {
+				return nil, ErrNoKeyForRole{Role: role}
+			}
+			return nil, err
+		}
+	} else if role == data.CanonicalRootRole {
+		// if this is root metadata, we have to get the keys from the root because they
+		// are certs
+		root := &data.Root{}
+		if err := json.Unmarshal(s.Signed, root); err != nil {
+			return nil, err
+		}
+		for _, pubKeyID := range root.Roles[data.CanonicalRootRole].KeyIDs {
+			if err := signed.Sign(cs, s, root.Keys[pubKeyID]); err != nil {
+				if _, ok := err.(signed.ErrNoKeys); ok {
+					return nil, ErrNoKeyForRole{Role: role}
+				}
+				return nil, err
+			}
+		}
+	} else {
+		pubKeyIDs := cs.ListKeys(role)
+		if len(pubKeyIDs) < 1 {
+			return nil, ErrNoKeyForRole{role}
+		}
+		for _, pubKeyID := range pubKeyIDs {
+			if err := signed.Sign(cs, s, cs.GetKey(pubKeyID)); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	metaBytes, err := json.Marshal(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return metaBytes, nil
+}
+
+// gets a Signed from the metadata store
+func signedFromStore(cache store.MetadataStore, role string) (*data.Signed, error) {
+	b, err := cache.GetMeta(role, maxSize)
+	if err != nil {
+		return nil, err
+	}
+
+	signed := &data.Signed{}
+	if err := json.Unmarshal(b, signed); err != nil {
+		return nil, err
+	}
+
+	return signed, nil
+}
+
+var delegatedRoles = []string{"targets/a", "targets/a/b"}
+
+// DefaultRoles are the defualt roles that NewMetadataSwizzler creates
+var DefaultRoles = append(data.BaseRoles, delegatedRoles...)
+
+// NewMetadataSwizzler creates a new tuf.Repo and generates metadata to fuzz.
+func NewMetadataSwizzler(gun string) (*MetadataSwizzler, error) {
```

Comment:
Ok, so my new plan:  move creating an empty repo with delegations to `testutils/repo.go`, move exporting all signed metadata to `tuf/tuf.go` (this way we can remove `client/helper.go`'s `serializeCanonicalRole`), and have swizzler take a `store.MetadataStore`.  It might be dependent upon #475 though (since `GetAllRoles` is super useful!)